cpp/tests/unary/unary_ops_test.cu

```diff
@@ -479,6 +529,61 @@ TEST_F(CastTimestamps, IsIdempotent) {
       timestamps_ns, *timestamps_ns_got);
 }
 
+template <typename T>
+struct CastTimestampsTyped : public cudf::test::BaseFixture {};
+
+TYPED_TEST_CASE(CastTimestampsTyped, cudf::test::TimestampTypes);
+
+// Return a list of timestamp type ids whose precision is greater than or equal
+// to the input type id
+std::vector<cudf::type_id> get_higher_precision_timestamp_type_ids(
+    cudf::type_id search) {
+  bool found = false;
+  std::vector<cudf::type_id> gte_ids{};
+  std::vector<cudf::type_id> timestamp_ids{
+      cudf::TIMESTAMP_DAYS, cudf::TIMESTAMP_SECONDS,
+      cudf::TIMESTAMP_MILLISECONDS, cudf::TIMESTAMP_MICROSECONDS,
+      cudf::TIMESTAMP_NANOSECONDS};
+  for (cudf::type_id type_id : timestamp_ids) {
+    if (found || (found = (type_id == search))) {
+      gte_ids.emplace_back(type_id);
+    }
+  }
+  return gte_ids;
+}
+
+// Test that all timestamps whose precision is >= to the TypeParam
+// down-casts appropriately to the lower-precision TypeParam
+TYPED_TEST(CastTimestampsTyped, DownCastingTruncatesValues) {
+  using T = TypeParam;
+  using namespace cudf::test;
+  auto dtype_exp = make_data_type<T>();
+  auto timestamps_exp = make_exp_timestamp_column(dtype_exp.id());
+  auto higher_precision_type_ids = get_higher_precision_timestamp_type_ids(
+      cudf::experimental::type_to_id<T>());
+  for (cudf::type_id higher_precision_type_id : higher_precision_type_ids) {
+    auto timestamps_src = make_exp_timestamp_column(higher_precision_type_id);
+    auto timestamps_got = cudf::experimental::cast(timestamps_src, dtype_exp);
+    expect_columns_equal(*timestamps_got, timestamps_exp);
+  }
+}
+
+// Specific test to ensure down-casting to days happens correctly
+TYPED_TEST(CastTimestampsTyped, DownCastingToDaysTruncatesValues) {
```

Comment:
Should `Truncates` be `Floors` in these test names?  Truncating is rounding towards zero (the epoch) which is the problematic behavior.  Downcasting needs to floor values rather than truncate them to round towards -Infinity.

Reply:
Yes, I agree that makes more sense. Changed in https://github.com/rapidsai/cudf/pull/3670/commits/137d5ceb0695cfd12907bfb020995a49f92cc852!